### ISSUE TITLE
Bug: `set_radio()` not working as expected with dynamic default value

### ISSUE THREAD

**Issue by ambadyanands** [codeigniter4/CodeIgniter4#6030]
### PHP Version

8.1

### CodeIgniter4 Version

4.1.9

### CodeIgniter4 Installation Method

Composer (using `codeigniter4/appstarter`)

### Which operating systems have you tested for this bug?

Linux

### Which server did you use?

cli

### Database

MariaDB 10.7.3

### What happened?

I tried to set the default value dynamically by comparing the radio button's value with the value already stored in db. But this breaks the functionality of `set_radio()` by setting the value from db as `checked` always.

### Steps to Reproduce

See https://github.com/codeigniter4/CodeIgniter4/issues/6030#issuecomment-1138556818

Consider the following scenario:

```
set_radio('someField', 'someValue', $valueFromDb === 'someValue')
set_radio('someField', 'someOtherValue', $valueFromDb === 'someOtherValue')
```

If the `$valueFromDb` is equal to `someOtherValue` and I click on the button with `someValue`, both radio buttons will be marked with `checked="checked"` (if two radio buttons have the checked property, the last button will be shown as checked for the user. But `inspect element` shows that both buttons have the checked property).

### Expected Output

On form validation error the last user-given input shows as checked instead of both or the one from db as checked.

### Anything else?

I think this is because `set_radio()` function checks for the `$default` value regardless of whether `$input` is non-empty. Instead it should only check for `$default` value when `$input` value is empty, this way `$input` value always gets precedence over the `$default` value.

I tried changing the following line:
 https://github.com/codeigniter4/CodeIgniter4/blob/202f41ad522ba1d414b9d9c35aba1cb0c156b781/system/Helpers/form_helper.php#L678

to:
```
if (empty($result) && (string) $input !== '0' && empty($input = $request->getPost($field)) && empty($input = old($field))) {
```

and it seems to be working fine.

Sorry in advance if I'm not using the `set_radio()` function correctly. I'd be pleased if there's a better solution, thanks.

**Comment by kenjis** [comment]
@ambadyanands Thank you for reporting.

But your code is not complete and does not work. So I cannot reproduce.

```php
set_radio('someField', 'someValue', $valueFromDb === 'someValue')
set_radio('someField', 'someOtherValue', $valueFromDb === 'someOtherValue')
```

We need minimum code but can reproduce the bug.
Thank you.

**Comment by ambadyanands** [comment]
Sorry, I thought it was obvious.

**Comment by ambadyanands** [comment]
Sorry, I thought it was obvious

**Comment by ambadyanands** [comment]
Sorry, the issue got closed accidentally.

Here's the code:
```
<input type="radio" name="someFiled" value="someValue" <?= set_radio('someFiled', 'someValue', $valueFromDb === 'someValue') ?> />
<input type="radio" name="someFiled" value="someOtherValue" <?= set_radio('someField', 'someOtherValue', $valueFromDb === 'someOtherValue') ?> />
```

where `$valueFromDb` is passed from controller and it can be assumed as `'someOtherValue'` as I mentioned above. Please let me know if I should provide any other details.

**Comment by kenjis** [comment]
It seems you use validation, but there is no validation code.

I executed the following code and radio button works fine.

```php
<?php
namespace App\Controllers;
class Home extends BaseController
{
    public function index()
    {
        helper('form');
        $valueFromDb = 'someOtherValue';
        return view('welcome_message', ['valueFromDb' => $valueFromDb]);
    }
}
```
```php
<!DOCTYPE html>
<html lang="en">
<body>

<input type="radio" name="someFiled" value="someValue" <?= set_radio('someFiled', 'someValue', $valueFromDb === 'someValue') ?> />
<input type="radio" name="someFiled" value="someOtherValue" <?= set_radio('someField', 'someOtherValue', $valueFromDb === 'someOtherValue') ?> />

</body>
</html>
```

**Comment by kenjis** [comment]
@ambadyanands All we need is minimum code that reproduces the problem.
I can't see your screen, and I don't want to guess your code. It is waste of time.

I hope you understand what I mean.
Thanks in advance.

**Comment by ambadyanands** [comment]
Okay, I think I didn't explain the issue properly (I made too many assumption, sorry about that). When the form loads for the first time, it will work fine. i.e. we see the value from db is checked as expected. The issue is when some form validation error occurs (like explained in the `Expected Output` section of the bug report), for eg. a required field is not filled and submitted. The following code should help to reproduce the issue.

```
<?php
namespace App\Controllers;
class Home extends BaseController
{
    public function index()
    {
        helper(['form', 'url']);
        $valueFromDb = '0';
        if (! $this->validate(['name' => 'required'])) {
            echo view('welcome_message', [
                'valueFromDb' => $valueFromDb,
                'validation' => $this->validator,
            ]);
        } else {
            echo 'Success';
        }
    }
}
```

```
<!DOCTYPE html>
<html lang="en">
<body>
    
<?= $validation->listErrors(); ?>
<?= form_open(base_url()); ?>
  <label for="name">Name</label>
  <input type="text" name="name"><br>
  <input type="radio" name="status" value="1" <?= set_radio('status', '1', $valueFromDb === '1') ?>>
  <label for="enabled">Enabled</label><br>
  <input type="radio" name="status" value="0" <?= set_radio('status', '0', $valueFromDb === '0') ?>>
  <label for="disabled">Disabled</label><br>
  <input type="submit" value="Submit">
<?= form_close(); ?> 

</body>
</html>
```

This will load the form with `Disabled` as checked by default. Now try clicking the `Enabled` button and submit the form without filing the text field. Instead of `Enabled` button, `Disabled` will be shown as checked. `Inspect element` shows that both buttons will have the `checked="checked"` attribute, but in the form only the last one with that attribute will be shown as checked for the user. What we expect is to have the `checked="checked"` property only for the `Enabled` button (i.e. the last user-given value should precede the default value). Hope that makes it clear, thanks.

**Comment by kenjis** [comment]
@ambadyanands Thank you for your cooperation.
I've confirm the bug. Good reporting!

**Comment by ambadyanands** [comment]
@kenjis Good catch. The trick is in checking the second button by default. When there are only two radio buttons and the first button is checked by default and one clicks on the second and submit it will show second as checked. But actually it was just browser showing the last one as checked when both have checked attribute and the bug will hide in plain sight! :)

**Comment by kenjis** [comment]
I sent a PR to fix this: #6037

**Comment by ambadyanands** [comment]
@kenjis Thanks, also the code looks more clean and easy to read now.

**Comment by kenjis** [comment]
This bug has been fixed in the `develop` branch,
and it would be included in v4.2.0.
